### PR TITLE
Pytorch implementation init

### PR DESCRIPTION
**Pull request by armenabnousi**
Adding ContentNet, DistortionNet, and CompressionNet implementation using the baseline weights.
 
At the moment, the features and labels from these sub-nets are being written into numpy's npy files. Once the implementation is complete, these should be fed into AggregationNet to generate the final results.

### CODE REVIEW

File: uvq_pytorch/utils/contentnet.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+# ContentNet specs
+DIM_LABEL_CONTENT = 3862
+
+
+class ContentNet(nn.Module):
+    def __init__(self, num_classes=DIM_LABEL_CONTENT, dropout=0.2):
+        super().__init__()
+        stochastic_depth_prob_step = 0.0125
+        stochastic_depth_prob = [x * stochastic_depth_prob_step for x in range(16)]
+        self.features = nn.Sequential(
```
Comment (yaohongwu-g) [inline uvq_pytorch/utils/contentnet.py:32]: Can we use the efficient net implementation from torchvision.models? http://docs/document/d/1-DgRdkPvt6LVJesD3bbW6rfbUJiHrgvXy-nIgRKhNpA?resourcekey=0-j_VtCkUxMZtoAYqGSplEEg&tab=t.0#heading=h.8n3y1cyqtsez

Comment (armenabnousi) [reply uvq_pytorch/utils/contentnet.py:32]: yes and no. The problem is pytorch (and torchvision's efficinetNet) don't use "same" padding in the convolutions. So the output sizes and as a result the final outputs are a bit different IIRC.
Inititally I had a different implementation that was using the efficientNet implementation but then looping through the architecture and changing all conv2d layers to Conv2dSamePadding custom layer that I have included in this code. I thought this version is cleaner.

Comment (yaohongwu-g) [reply uvq_pytorch/utils/contentnet.py:32]: You are right. We saw this issue before and we used the alternative approach you mentioned (loop + replace conv layer). I think it is fine to explicitly list out all the layers, as long as the model output is similar with TF.

Comment (armenabnousi) [reply uvq_pytorch/utils/contentnet.py:32]: The outputs of the two approaches are the same. Let me know if you prefer to change this and I'll prepare the other version.

Comment (yaohongwu-g) [reply uvq_pytorch/utils/contentnet.py:32]: I am OK with the current implementation (perhaps you can add a bit more comment in the code to explain the rationale). We can change that later if needed.

I personally slightly prefer the alternative approach (based on torchvision's efficientnet). The Conv2dSamePadding layer is only needed if you want to copy weights from our TF model. If in the future, you want to retrain everything from scratch, then you probably do not need the Conv2dSamePadding layer.

@yilinwang01 FYI

File: uvq_pytorch/utils/distortionnet.py [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+                video.shape[0],
+                self.feature_height,
+                self.feature_width,
+                self.feature_channels,
+            ),
+            np.float32,
+        )
+
+        if self.depth == 1:
+            # TODO: this is for distortionnet; for compressionnet keep only the "else" part.
```
Comment (yaohongwu-g) [inline uvq_pytorch/utils/distortionnet.py:304]: Maybe we can remove this comment and also the "else" part.

Maybe add an assertion somewhere to make sure depth is always 1 for now.

File: uvq_pytorch/utils/distortionnet.py [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+                    self.patch_width,
+                ),
+                np.float32,
+            )
+
+        for k in range(video.shape[0]):
+            for j in range(self.num_patches_y):
+                for i in range(self.num_patches_x):
+                    if self.depth == 1:
+                        # TODO: this would work for distortionnet, it's not used here and should be cleaned up; keep only the "else" part
```
Comment (yaohongwu-g) [inline uvq_pytorch/utils/distortionnet.py:326]: Same here. Remove this comment and remove the "else" part (the "if" is also redundant if there is only 1 branch).

File: uvq_pytorch/utils/compressionnet.py [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+                video.shape[0],
+                self.feature_height,
+                self.feature_width,
+                self.feature_channels,
+            ),
+            np.float32,
+        )
+
+        if self.depth == 1:
+            # TODO: this is for distortionnet; for compressionnet keep only the "else" part.
```
Comment (yaohongwu-g) [inline uvq_pytorch/utils/compressionnet.py:160]: Please clean this up.

File: uvq_pytorch/utils/compressionnet.py [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+                    self.patch_width,
+                ),
+                np.float32,
+            )
+
+        for k in range(video.shape[0]):
+            for j in range(self.num_patches_y):
+                for i in range(self.num_patches_x):
+                    if self.depth == 1:
+                        # TODO: this would work for distortionnet, it's not used here and should be cleaned up; keep only the "else" part
```
Comment (yaohongwu-g) [inline uvq_pytorch/utils/compressionnet.py:182]: Same here.

File: uvq_pytorch/utils/contentnet.py [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+    def infer_from_input_video(self, video_filename, video_length, transpose=False):
+        video, video_resized = VideoReader.load_video(
+            video_filename, video_length, transpose
+        )
+        video_features, video_labels = self.get_labels_and_features_for_all_frames(
+            video_resized
+        )
+        return video_features, video_labels
+
+    def get_labels_and_features_for_all_frames(self, video):
```
Comment (yaohongwu-g) [inline uvq_pytorch/utils/contentnet.py:135]: I would recommend adding a docstring and type annotation for this function (and any other public functions in this file or other files).

For example:
```
def get_labels_and_features_for_all_frames(self, video: np.ndarray) -> tuple[np.ndarray, np.ndarray]:
"""
Gets predicted labels and features for all frames of a video.

Args:
  video: parsed video 5d array with shape (num_seconds, fps, height, width, 3) where each value is in [-1, 1] .

Returns:
  feature: ...
  label: ...

"""
```

Reference: https://google.github.io/styleguide/pyguide.html#383-functions-and-methods

**Review by yaohongwu-g — CHANGES_REQUESTED** [review]
Armen, thanks for making these changes (and sorry for my late reply).

Left some comments, mostly about the style / readability.

Also, could you sign the Contributor License Agreement (CLA) when you have a chance? Thanks!

File: uvq_pytorch/readme.md [hunk 7]
```diff
@@ -0,0 +1,9 @@
+To test the pytorch implementation, `cd` into the `uvq_pytorch` directory.
+Download an example video. E.g. [Gaming_1080P-0ce6_orig.mp4](https://storage.googleapis.com/ugc-dataset/vp9_compressed_videos/Gaming_1080P-0ce6_orig.mp4)
+Then run:
+```
+python inference.py Gaming_1080P-0ce6_orig.mp4 20
+```
+first argument being the name of the file and the second one the length of the video.
+
+This will generate 6 files corresponding to ContentNet, CompressionNet, DistortionNet features and labels.
```
[the remark this answers is not in the document]
Comment (armenabnousi) [reply uvq_pytorch/readme.md]: yes, fixed the readme

File: uvq_pytorch/inference.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+            )
+        )
+        results = self.aggregationnet.predict(
+            compression_features, content_features, distortion_features
+        )
+        print(results)
+        return results
+
+    def load_video(self, video_filename, video_length, transpose=False):
+        video_resized1, video_resized2 = VideoReader.load_video(
```
[the remark this answers is not in the document]
Comment (armenabnousi) [reply uvq_pytorch/inference.py]: moved out the video reader to uvq/utils and made it common between tf and pytorch. Because this one requires (channel, height, width), while tf requires (height, width, channel), I have added a transpose here

File: uvq_pytorch/utils/contentnet.py [hunk 9]
```diff
@@ -0,0 +1,154 @@
+import numpy as np
+import pandas as pd
+import torch
+
+from torch import nn, Tensor
+
+from .custom_nn_layers import (
```
[the remark this answers is not in the document]
Comment (armenabnousi) [reply uvq_pytorch/utils/contentnet.py]: should be according to the guide now

**Comment by armenabnousi** [comment]
Thanks for the check Yaohong. I have made the changes according to the comments. Sorry about the late response (neurIPS + holidays + annual reviews slowed things down).

For the CLA agreement - I'm waiting on the legal, I think I should hear back something next week once they're back from a pto.

File: uvq_pytorch/inference.py [hunk 10]
```diff
@@ -44,21 +60,65 @@ def infer(self, video_filename, video_length, transpose=False):
...
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "video_filename",
+        type=str,
+        help="Path to the video file",
+    )
+    parser.add_argument(
+        "video_length",
+        type=int,
+        help="Length of the video in frames",
```
Comment (yaohongwu-g) [inline uvq_pytorch/inference.py:107]: video_length is in seconds (instead of frames), right?

File: uvq_pytorch/utils/aggregationnet.py [hunk 11]
```diff
@@ -1,28 +1,47 @@
+import typing
```
Comment (yaohongwu-g) [inline uvq_pytorch/utils/aggregationnet.py:1]: nit: you can import symbols directly from typing (e.g., `from typing import Iterator`), see https://google.github.io/styleguide/pyguide.html#31912-imports-for-typing (this is a exemption of import rule https://google.github.io/styleguide/pyguide.html#2241-exemptions).

**Review by yaohongwu-g — CHANGES_REQUESTED** [review]
Thanks for making the changes. Overall it looks good to me. Left a few minor comments.

**Comment by armenabnousi** [comment]
thanks for the suggestions. They're resolved now. I have been told we should expect cla agreement signature in the next two weeks

**Comment by yaohongwu-g** [comment]
> thanks for the suggestions. They're resolved now. I have been told we should expect cla agreement signature in the next two weeks

Hi Armen, is there any update on signing CLA?

**Comment by armenabnousi** [comment]
Sorry about the delay. New process. They said ~1 week on Tuesday.

**Comment by armenabnousi** [comment]
I have signed the CLA and amended the last commit to match the email address associated with the CLA.

**Comment by armenabnousi** [comment]
> Before we can accept the commits, could you add the following copyright info (to adhere to our release process and be consistent with existing files) in the header of each .py file you add?

Thanks! Added!

**Comment by yaohongwu-g** [comment]
Thanks Armen. Accepting the PR.
